### PR TITLE
Add tuition and fees deduction

### LINKED ISSUE #4035: Tuition and fees deduction
IRS Form 8917

Data available in the PUF (~$7 billion)

### PR DESCRIPTION
**Pull request by hua7450**
Fixes #4035

### CODE REVIEW

**Comment by PavelMakarchuk** [comment]
@hua7450 new PR?

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        adjusted_gross_income = tax_unit("adjusted_gross_income", period)
+        # married filing separatly are not eligible for this deduction
+        filing_status = tax_unit("filing_status", period)
+        not_separate = filing_status != filing_status.possible_values.SEPARATE
+        # can't caim this deduction if the household has American oppportunity or lifetime learning credit
+        american_opportunity_credit = tax_unit(
+            "american_opportunity_credit", period
+        )
+        lifetime_learning_credit = tax_unit("lifetime_learning_credit", period)
+        eligible = (
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:29]:
```suggestion
        has_take_the_american_opp_or_life_learn_credits = (
```

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            american_opportunity_credit + lifetime_learning_credit
+        ) == 0
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:31]:
```suggestion
        ) > 0
```

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        p = parameters(period).gov.irs.deductions.tuition_and_fees
+        joint = filing_status == filing_status.possible_values.JOINT
+        cap = where(
+            joint,
+            p.joint.calc(adjusted_gross_income),
+            p.non_joint.calc(adjusted_gross_income),
+        )
+        return min_(qualified_tuition_expenses * eligible * not_separate, cap)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:39]:
```suggestion
        return capped_expenses * ~eligible * not_separate
```

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        eligible = (
+            american_opportunity_credit + lifetime_learning_credit
+        ) == 0
+        p = parameters(period).gov.irs.deductions.tuition_and_fees
+        joint = filing_status == filing_status.possible_values.JOINT
+        cap = where(
+            joint,
+            p.joint.calc(adjusted_gross_income),
+            p.non_joint.calc(adjusted_gross_income),
+        )
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:38]:
```suggestion
        )
        capped_expenses = min_(qualified_tuition_expenses, cap)

```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        # Law was repealed in 2021
+        "https://irc.bloombergtax.com/public/uscode/doc/irc/section_222"
+    )
+
+    def formula(tax_unit, period, parameters):
+        qualified_tuition_expenses = add(
+            tax_unit, period, ["qualified_tuition_expenses"]
+        )
+        adjusted_gross_income = tax_unit("adjusted_gross_income", period)
+        # married filing separatly are not eligible for this deduction
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:21]:
```suggestion
        # Married filing separately are not eligible for this deduction.
```

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        filing_status = tax_unit("filing_status", period)
+        separate = filing_status == filing_status.possible_values.SEPARATE
+        # can't caim this deduction if the household has American oppportunity or lifetime learning credit
+        american_opportunity_credit = tax_unit(
+            "american_opportunity_credit", period
+        )
+        lifetime_learning_credit = tax_unit("lifetime_learning_credit", period)
+        has_take_the_american_opp_or_life_learn_credits = (
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:29]:
```suggestion
        has_taken_the_american_opp_or_life_learn_credits = (
```

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    def formula(tax_unit, period, parameters):
+        qualified_tuition_expenses = add(
+            tax_unit, period, ["qualified_tuition_expenses"]
+        )
+        adjusted_gross_income = tax_unit("adjusted_gross_income", period)
+        # married filing separatly are not eligible for this deduction
+        filing_status = tax_unit("filing_status", period)
+        separate = filing_status == filing_status.possible_values.SEPARATE
+        # can't caim this deduction if the household has American oppportunity or lifetime learning credit
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:24]:
```suggestion
        # Can't claim this deduction if the household has taken the
        # American Opportunity Credit or Lifetime Learning Credit.
```

File: policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        american_opportunity_credit = tax_unit(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.py:25]: Please make a separate variable `tuition_and_fees_deduction_eligible` that looks like this
```suggestion
        aoc_llc = add(tax_unit, period, ["american_opportunity_credit", "lifetime_learning_credit"])
        return ~separate & (aoc_llc == 0)
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
Thanks. For posterity, the 2021 Consolidated Appropriations Act repealed this, shifting it to a higher Lifetime Learning Credit

File: policyengine_us/tests/policy/baseline/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.yaml [hunk 9]
```diff
@@ -0,0 +1,88 @@
+- name: Test case 1 #Separate filing, agi less than 65000, qualified tuition expenses less than 4000
+  period: 2020
+  input:
+    filing_status: SEPARATE
+    qualified_tuition_expenses: 2_500
+    adjusted_gross_income: 55_000
+    american_opportunity_credit: 0
```
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/irs/income/taxable_income/deductions/tuition_and_fees_deduction.yaml:7]: Use eligibility variable as input instead

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**